File: src/nodejs/nestjs/src/controllers/database.controller.ts
import { Controller, Get, Query } from '@nestjs/common';
import { DatabaseService } from '../services/database.service';

@Controller('db')
export class DatabaseController {
  constructor(private readonly databaseService: DatabaseService) {}

  @Get('simple')
  async getSimple(@Query('id') id: string) {
    if (!id) {
      return {
        error: 'Bad Request',
        message: 'id parameter is required',
      };
    }

    const userId = parseInt(id);
    if (isNaN(userId)) {
      return {
        error: 'Bad Request',
        message: 'id must be a number',
      };
    }

    const user = await this.databaseService.findUserById(userId);

    if (!user) {
      return {
        error: 'Not Found',
        message: 'User not found',
      };
    }

    // The contract returns the user object itself, not an envelope, and the
    // normative SQL already aliases the columns to the contract names.
    return user;
  }

  // The previous version answered {orders, summary: {...}} and took a userId
  // query parameter no other implementation accepted, which also meant it was
  // parsing an argument the others were not.
  // See contracts/rest/canonical-payloads.md.
  @Get('complex')
  async getComplex(@Query('days') days: string = '30') {
    const parsedDays = parseInt(days);

    if (isNaN(parsedDays) || parsedDays <= 0 || parsedDays > 365) {
      return {
        error: 'Bad Request',
        message: 'days must be between 1 and 365',
      };
    }

    const data = await this.databaseService.findComplexOrders(parsedDays);

    return {
      periodDays: parsedDays,
      totalUsers: data.length,
      data,
    };
  }
}

File: src/graalvm/gspring/src/main/java/com/benchmark/controller/DatabaseController.java
package com.benchmark.controller;

import com.benchmark.model.User;
import com.benchmark.model.UserStats;
import com.benchmark.service.DatabaseService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
public class DatabaseController {
    private final DatabaseService databaseService;

    @Autowired
    public DatabaseController(DatabaseService databaseService) {
        this.databaseService = databaseService;
    }

    @GetMapping("/db/simple")
    public Object dbSimple(@RequestParam(required = false) String id) {
        Integer userId = 1;
        if (id != null && !id.isEmpty()) {
            try {
                userId = Integer.parseInt(id);
            } catch (NumberFormatException e) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", "Bad Request");
                error.put("message", "id must be a number");
                return error;
            }
        }

        Optional<User> user = databaseService.getUserById(userId);
        if (user.isEmpty()) {
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Not Found");
            error.put("message", "User with id " + userId + " not found");
            return error;
        }

        return user.get();
    }

    @GetMapping("/db/complex")
    public Map<String, Object> dbComplex(@RequestParam(required = false) String days) {
        Integer daysInt = 30;
        if (days != null && !days.isEmpty()) {
            try {
                daysInt = Integer.parseInt(days);
                if (daysInt <= 0 || daysInt > 365) {
                    Map<String, Object> error = new HashMap<>();
                    error.put("error", "Bad Request");
                    error.put("message", "days must be between 1 and 365");
                    return error;
                }
            } catch (NumberFormatException e) {
                Map<String, Object> error = new HashMap<>();
                error.put("error", "Bad Request");
                error.put("message", "days must be a number");
                return error;
            }
        }

        List<UserStats> stats = databaseService.getUserStats(daysInt);

        Map<String, Object> response = new HashMap<>();
        response.put("period_days", daysInt);
        response.put("total_users", stats.size());
        response.put("data", stats);
        return response;
    }
}
